src/com/ms/silverking/cloud/dht/daemon/storage/convergence/ConvergenceController2.java

```diff
@@ -65,495 +69,500 @@
+    displayOutstandingRequests(outstandingChecksumTreeRequests, "outstandingChecksumTreeRequests");
+    displayOutstandingRequests(outstandingSyncRetrievalRequests, "outstandingSyncRetrievalRequests");
+    Log.warningf("queuedChecksumTreeRequests:\t%d\n", queuedChecksumTreeRequests.size());
+    Log.warning(
+        String.format("convergenceControllers %d\tactiveConvergenceControllers %d\n", convergenceControllers.size(),
+            activeConvergenceControllers.size()));
+  }
+
+  //////////////////////////////////////////////////////////////////////
+  // Generate requests for convergence
+
+  public void setChainNext(ConvergenceController2 chainNext) {
+    chainNextLock.lock();
+    try {
+      assert this.chainNext == null;
+      this.chainNext = chainNext;
+      Log.warning(String.format("%s setting chainNext %s", myUUID, chainNext.myUUID));
+    } finally {
+      chainNextLock.unlock();
     }
+  }
 
-    private void signalChainNext() {
-        chainNextLock.lock();
-        try {
-            if (chainNext != null) {
-                if (!chainNextSignaled) {
-                    Log.warning(String.format("%s signaling chainNext %s", myUUID, chainNext.myUUID));
-                    chainNextSignaled = true;
-                    chainNext.startConvergence();
-                } else {
-                    Log.warning(String.format("%s chainNext already signaled", myUUID));
-                }
-            } else {
-                Log.warning(String.format("%s chainNext is null", myUUID));
-            }
-        } finally {
-            chainNextLock.unlock();
+  private void signalChainNext() {
+    chainNextLock.lock();
+    try {
+      if (chainNext != null) {
+        if (!chainNextSignaled) {
+          Log.warning(String.format("%s signaling chainNext %s", myUUID, chainNext.myUUID));
+          chainNextSignaled = true;
+          chainNext.startConvergence();
+        } else {
+          Log.warning(String.format("%s chainNext already signaled", myUUID));
         }
+      } else {
+        Log.warning(String.format("%s chainNext is null", myUUID));
+      }
+    } finally {
+      chainNextLock.unlock();
     }
-    
-    public void startConvergence() {
-        convergencePauseMutex.lock();
+  }
+
+  public void startConvergence() {
+    convergencePauseMutex.lock();
+    try {
+      if (convergencePaused) {
         try {
-            if (convergencePaused) {
-                try {
-                    Log.warning("Storing paused signalled convergence controller ", myUUID);
-                    pausedSignalledControllers.put(this);
-                } catch (InterruptedException e) {
-                };
-                return;
-            }
-        } finally {
-            convergencePauseMutex.unlock();
+          Log.warning("Storing paused signalled convergence controller ", myUUID);
+          pausedSignalledControllers.put(this);
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)